src/renderer/base/renderer.cpp

```diff
@@ -16,29 +16,30 @@ static constexpr auto maxRetriesForRenderEngine = 3;
 // The renderer will wait this number of milliseconds * how many tries have elapsed before trying again.
 static constexpr auto renderBackoffBaseTimeMilliseconds{ 150 };
 
+#define FOREACH_ENGINE(var)   \
+    for (auto var : _engines) \
+        if (!var)             \
+            break;            \
+        else

```

Comment:
Is it worth adding a comment here explaining that `_engines` is kind of a null-terminated list, just so it's clearer why we're breaking out of the loop on `!var`? And on that note, would it be worthwhile making the `_engines` array one entry larger so you can always assume it's null-terminated, and not also have to check the size?

Reply:
I've tried that today. Counterintuitively this increases binary size by 500 byte and makes no difference for performance. So I'll keep it this way for now. for-range loops seem to be easier to optimize for compilers...